cf/test/test_Data.py

```diff
@@ -977,103 +977,98 @@ def test_Data_cached_arithmetic_units(self):
+
+        # Operation with equivalent but non-equal units:
+        d_np = np.array([[1, 2], [3, 4]])
+        e_np = np.array([[5.0, 6.0]])
+        d = cf.Data(d_np, "km")
+        e = cf.Data(e_np, "metre")
+        f_np = np.concatenate((d_np, e_np / 1000))  # /1000 for unit conversion
+        f = cf.Data.concatenate([d, e])  # (and try a list input)
+
+        self.assertEqual(f.shape, f_np.shape)
+        self.assertTrue((f.array == f_np).all())
+
+        # Check axes equivalency:
+        self.assertTrue(f.equals(cf.Data.concatenate((d, e), axis=-2)))
+
+        # Non-default axis specification:
+        e_np = np.array([[5.0], [6.0]])  # for compatible shapes with axis=1
+        e = cf.Data(e_np, "metre")
+        f_np = np.concatenate((d_np, e_np / 1000), axis=1)
+        f = cf.Data.concatenate((d, e), axis=1)
+
+        self.assertEqual(f.shape, f_np.shape)
+        self.assertTrue((f.array == f_np).all())
+
+        # Operation with every data item in sequence being a scalar:
+        d_np = np.array(1)
+        e_np = np.array(50.0)
+        d = cf.Data(d_np, "km")
+        e = cf.Data(e_np, "metre")
+
+        # Note can't use the following (to compute answer):
+        #     f_np = np.concatenate([d_np, e_np])
+        # here since we have different behaviour to NumPy w.r.t scalars, where
+        # NumPy would error for the above with:
+        #     ValueError: zero-dimensional arrays cannot be concatenated
+        f_answer = np.array([d_np, e_np / 1000])  # /1000 for unit conversion
+        f = cf.Data.concatenate((d, e))
+
+        self.assertEqual(f.shape, f_answer.shape)
+        self.assertTrue((f.array == f_answer).all())
+
+        # Operation with some scalar and some non-scalar data in the sequence:
+        e_np = np.array([50.0, 75.0])
+        e = cf.Data(e_np, "metre")
+
+        # As per above comment, can't use np.concatenate to compute
+        f_answer = np.array([1.0, 0.05, 0.075])  # manual /1000 unit conversion
+        f = cf.Data.concatenate((d, e))
+
+        self.assertEqual(f.shape, f_answer.shape)
+        self.assertTrue((f.array == f_answer).all())
+
+        # Check the cyclicity of axes is correct after concatenate...
+        d_np = np.arange(16).reshape(4, 4)
+        d = cf.Data(d_np, "seconds")
+        d.cyclic([0, 1])  # notably make both axes cyclic here
+        e_np = d_np.copy()
+        e = cf.Data(e_np, "seconds")
+
+        f_np = np.concatenate((d_np, e_np))
+
+        # ...when joining along axis=0 (the default)
+        self.assertEqual(d.cyclic(), {0, 1})
+        f = cf.Data.concatenate([d, e])
+        self.assertEqual(f.cyclic(), {1})
+
+        self.assertEqual(f.shape, f_np.shape)
+        self.assertTrue((f.array == f_np).all())
+
+        # ...when joining along axis=1
+        f_np = np.concatenate((d_np, e_np), axis=1)
+
+        self.assertEqual(d.cyclic(), {0, 1})
+        f = cf.Data.concatenate([d, e], axis=1)
+        self.assertEqual(f.cyclic(), {0})
```

Comment:
@davidhassell this part of the test method should capture the behaviour I believe we need with regards to adjusting the cyclicity of an axis appropriately if it was used as the join axis for a concatenation. When reviewing, please check in particular that this is what you envisioned, also. (Obviously, the behaviour I have implemented for this PR makes this updated test pass so this is what is happening as-is.)

Reply:
Looks good to me (see comments on cyclic bit of `concatenate` code)